plugins/BEdita/API/tests/TestCase/Middleware/TokenMiddlewareTest.php

```diff
@@ -0,0 +1,231 @@
+                [
+                    'app' => ['id' => 1],
+                    'payload' => null,
+                ],
+                [],
+                [
+                    'api_key' => API_KEY,
+                ],
+            ],
+            'no token' => [
+                [
+                    'payload' => null,
+                ],
+                [
+                ],
+            ],
+        ];
+    }
+
+    /**
+     * Test middleware invocation.
+     *
+     * @return void
+     *
+     * @dataProvider invokeProvider
+     *
+     * @covers ::__invoke()
+     * @covers ::readApplication()
+     * @covers ::getToken()
+     * @covers ::applicationFromApiKey()
+     * @covers ::fetchApiKey()
+     * @covers ::verifyClientCredentials()
+     */
+    public function testInvoke($expected, array $server, ?array $query = null, bool $blockAnonymous = false): void
+    {
+        if ($expected instanceof \Exception) {
+            $this->expectException(get_class($expected));
+            $this->expectExceptionMessage($expected->getMessage());
+        }
+        Configure::write('Security.blockAnonymousApps', $blockAnonymous);
+
+        $request = ServerRequestFactory::fromGlobals($server, $query);
+        $response = new Response();
+        $next = function ($req, $res) {
+            return $req;
+        };
+        $middleware = new TokenMiddleware();
+        /** @var \Zend\Diactoros\ServerRequest $result */
+        $result = $middleware($request, $response, $next);
+
+        $app = CurrentApplication::getApplication();
+        if (empty($expected['app'])) {
+            static::assertNull($app);
+        } else {
+            $id = Hash::get($expected, 'app.id');
+            static::assertEquals($app->id, $id);
+        }
+
+        $payload = $result->getAttribute(TokenMiddleware::PAYLOAD_REQUEST_ATTRIBUTE, null);
+        $expectedPayload = Hash::get($expected, 'payload');
+        static::assertEquals($expectedPayload, $payload);
+    }
+
+    /**
+     * Test default behavior on missing 'Security.blockAnonymousApps' key
+     *
+     * @return void
+     *
+     * @covers ::fetchApiKey()
+     * @covers ::verifyClientCredentials()
+     */
+    public function testGetApplicationDefault()
+    {
+        $this->expectException(ForbiddenException::class);
+        $this->expectExceptionMessage('Missing API key');
+
+        Configure::delete('Security.blockAnonymousApps');
+
+        $middleware = new TokenMiddleware();
+        $middleware(
+            ServerRequestFactory::fromGlobals(),
+            new Response(),
+            null
+        );
+    }
+
+    /**
+     * Test default behavior on `client_credentials` request
+     *
+     * @return void
+     *
+     * @covers ::fetchApiKey()
+     * @covers ::verifyClientCredentials()
+     */
+    public function testClientCredentialsRequest()
+    {
+        CurrentApplication::setApplication(null);
+        Configure::delete('Security.blockAnonymousApps');
```

Comment:
Do we need to restore previously value of `'Security.blockAnonymousApps'` at the end of test?